File: electron/vault/ignore.ts
import fs from 'node:fs';
import path from 'node:path';

const IGNORE_FILE = '.thoughtstackignore';

/**
 * Reads ignore patterns from <vaultPath>/.thoughtstackignore.
 * Lines starting with # and blank lines are skipped.
 */
export function readIgnorePatterns(vaultPath: string): string[] {
  const filePath = path.join(vaultPath, IGNORE_FILE);
  if (!fs.existsSync(filePath)) return [];

  return fs
    .readFileSync(filePath, 'utf-8')
    .split('\n')
    .map((l) => l.trim())
    .filter((l) => l.length > 0 && !l.startsWith('#'));
}

/**
 * Appends a new pattern to .thoughtstackignore.
 * Creates the file if it doesn't exist. Skips duplicates.
 */
export function addIgnorePattern(vaultPath: string, pattern: string): void {
  const filePath = path.join(vaultPath, IGNORE_FILE);
  const existing = readIgnorePatterns(vaultPath);

  if (existing.includes(pattern)) return;

  const line = (fs.existsSync(filePath) ? '\n' : '') + pattern + '\n';
  fs.appendFileSync(filePath, line, 'utf-8');
}

/**
 * Returns true if the given relative path matches any ignore pattern.
 * A pattern matches if the path equals it or starts with it followed by '/'.
 */
export function isIgnored(patterns: string[], relativePath: string): boolean {
  const normalized = relativePath.replace(/\\/g, '/');
  return patterns.some(
    (p) => normalized === p || normalized.startsWith(p + '/')
  );
}

File: src-tauri/src/vault/ignore.rs
use std::fs;
use std::path::Path;

const IGNORE_FILE: &str = ".thoughtstackignore";

pub fn read_ignore_patterns(vault_path: &Path) -> Vec<String> {
    let file_path = vault_path.join(IGNORE_FILE);
    match fs::read_to_string(&file_path) {
        Ok(content) => content
            .lines()
            .map(|l| l.trim().to_string())
            .filter(|l| !l.is_empty() && !l.starts_with('#'))
            .collect(),
        Err(_) => Vec::new(),
    }
}

pub fn add_ignore_pattern(vault_path: &Path, pattern: &str) -> Result<(), String> {
    let file_path = vault_path.join(IGNORE_FILE);
    let existing = read_ignore_patterns(vault_path);
    if existing.iter().any(|p| p == pattern) {
        return Ok(());
    }
    let prefix = if file_path.exists() { "\n" } else { "" };
    let line = format!("{}{}\n", prefix, pattern);
    fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(&file_path)
        .and_then(|mut f| {
            use std::io::Write;
            f.write_all(line.as_bytes())
        })
        .map_err(|e| e.to_string())
}

pub fn is_ignored(patterns: &[String], relative_path: &str) -> bool {
    let normalized = relative_path.replace('\\', "/");
    patterns
        .iter()
        .any(|p| normalized == *p || normalized.starts_with(&format!("{}/", p)))
}
